src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/WorkspaceComponent.scala

```diff
@@ -126,6 +126,45 @@ trait WorkspaceComponent {
       DBIO.seq(insertActions.toSeq:_*)
     }
 
+    private def upsertAttributes(workspace: Workspace) = {
+      val workspaceId = UUID.fromString(workspace.workspaceId)
+
+      def createWorkspaceAttributeTempTable() = {
+        sql"""create temporary table if not exists WORKSPACE_ATTRIBUTE_TEMP (id bigint(20) unsigned NOT NULL AUTO_INCREMENT primary key, name text NOT NULL, value_string text, value_number double DEFAULT NULL, value_boolean bit(1) DEFAULT NULL, value_entity_ref bigint(20) unsigned DEFAULT NULL, list_index int(11) DEFAULT NULL, owner_id binary(16) NOT NULL)""".as[Int]
+      }
+
+      def clearWorkspaceAttributeTempTable() = {
+        sql"""delete from WORKSPACE_ATTRIBUTE_TEMP""".as[Int]
+      }
+
+      def insertTempAttributes(): ReadWriteAction[Unit] = {
+        val entityRefs = workspace.attributes.filter(_._2.isInstanceOf[AttributeEntityReference]).map(_._2.asInstanceOf[AttributeEntityReference]).toSeq
+        val entityRefLists = workspace.attributes.filter(_._2.isInstanceOf[AttributeEntityReferenceList]).map(_._2.asInstanceOf[AttributeEntityReferenceList]).toSeq
+
+        val entitiesToLookup = (entityRefs ++ entityRefLists.map(_.list).flatten)
+
+        entityQuery.lookupEntitiesByNames(workspaceId, entitiesToLookup) flatMap { entityRecords =>
+          val entityIdsByRef = entityRecords.map(rec => AttributeEntityReference(rec.entityType, rec.name) -> rec.id).toMap
+          workspaceAttributeTempQuery.batchInsertAttributes(workspace.attributes.map(attr => workspaceAttributeTempQuery.marshalAttribute(workspaceId, attr._1, attr._2, entityIdsByRef)).flatten.toSeq)
+        }
+      }
+
+      object alterMasterActions extends RawSqlQuery {
+        val driver: JdbcDriver = WorkspaceComponent.this.driver
+        //if you drop the $workspaceId clause at the end it just may delete everything in the attributes table :)
+        val deleteFromMasterAction = sql"""delete ea from WORKSPACE_ATTRIBUTE ea left join WORKSPACE_ATTRIBUTE_TEMP ta on (ea.name,ea.owner_id,ea.list_index)<=>(ta.name,ta.owner_id,ta.list_index) where ta.owner_id is null and ea.owner_id=$workspaceId""".as[Int]
```

Comment:
this will delete from WORKSPACE_ATTRIBUTE any attribute that doesn't exist in WORKSPACE_ATTRIBUTE_TEMP but exists in WORKSPACE_ATTRIBUTE